accession-commons-mongodb/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/mongodb/repository/BasicMongoDbAccessionedCustomRepositoryImpl.java

```diff
@@ -0,0 +1,119 @@
+/*
+ *
+ * Copyright 2018 EMBL - European Bioinformatics Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.ebi.ampt2d.commons.accession.persistence.mongodb.repository;
+
+import com.mongodb.BulkWriteError;
+import org.springframework.data.mongodb.BulkOperationException;
+import org.springframework.data.mongodb.core.BulkOperations;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import uk.ac.ebi.ampt2d.commons.accession.core.SaveResponse;
+import uk.ac.ebi.ampt2d.commons.accession.persistence.IAccessionedObjectCustomRepository;
+import uk.ac.ebi.ampt2d.commons.accession.persistence.mongodb.document.AccessionedDocument;
+
+import java.io.Serializable;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class BasicMongoDbAccessionedCustomRepositoryImpl<
+        ACCESSION extends Serializable,
+        DOCUMENT extends AccessionedDocument<ACCESSION>>
+        implements IAccessionedObjectCustomRepository<ACCESSION, DOCUMENT> {
+
+    private final Class<DOCUMENT> clazz;
+    private final MongoTemplate mongoTemplate;
+
+    public BasicMongoDbAccessionedCustomRepositoryImpl(Class<DOCUMENT> clazz, MongoTemplate mongoTemplate) {
+        this.clazz = clazz;
+        this.mongoTemplate = mongoTemplate;
+    }
+
+    @Override
+    public SaveResponse<ACCESSION> insert(List<DOCUMENT> documents) {
+        checkHashUniqueness(documents);
+        putAuditSaveDate(documents);
+        final BulkOperations insert = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, clazz)
+                .insert(new ArrayList<>(documents));
+        final Set<String> duplicatedHash = new HashSet<>();
+
+        try {
+            insert.execute();
+        } catch (BulkOperationException e) {
+            e.getErrors().forEach(error -> {
+                String duplicatedId = parseIdDuplicateKey(error).orElseThrow(() -> e);
+                duplicatedHash.add(duplicatedId);
```

Comment:
What if it is another type of bulk error (you know Mongo :roll_eyes: )? We should report it some way.

Reply:
If it is another error that it is not a duplicateKey the error is thrown up, system failure.